File: src/main/java/com/donrobo/fpbg/blueprint/building/Splitter.java
package com.donrobo.fpbg.blueprint.building;

import com.donrobo.fpbg.blueprint.Direction;
import net.sf.json.JSONObject;

import static com.donrobo.fpbg.blueprint.Direction.*;

public class Splitter extends AbstractBuilding {

    private final Direction direction;

    public Splitter(int x, int y, Direction direction) {
        super(x, y);
        this.direction = direction;
    }

    @Override
    public int getWidth() {
        return direction == UP || direction == DOWN ? 2 : 1;
    }

    @Override
    public int getHeight() {
        return direction == UP || direction == DOWN ? 1 : 2;
    }

    @Override
    public Building move(int x, int y) {
        return new Splitter(getX() + x, getY() + y, direction);
    }

    @Override
    public String getName() {
        return "splitter";
    }

    @Override
    public char getVisualizationCharacter() {
        return 'S';
    }

    @Override
    protected double getBlueprintXOffset() {
        return direction == UP || direction == DOWN ? 0.5 : 0;
    }

    @Override
    protected double getBlueprintYOffset() {
        return direction == LEFT || direction == RIGHT ? 0.5 : 0;
    }

    @Override
    protected void addCustomPropertiesToJson(JSONObject json) {
        json.put("direction", direction.reverseDirection().getDirectionValue());
    }
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/building/Splitter.kt
package com.donrobo.fpbg.blueprint.building

import com.donrobo.fpbg.blueprint.Direction
import com.donrobo.fpbg.blueprint.Direction.*
import com.donrobo.fpbg.data.Int2
import net.sf.json.JSONObject

class Splitter(x: Int, y: Int, val direction: Direction) : AbstractBuilding(x, y) {

    override val width: Int
        get() = if (direction === UP || direction === DOWN) 2 else 1

    override val height: Int
        get() = if (direction === UP || direction === DOWN) 1 else 2

    override fun move(x: Int, y: Int): Building {
        return Splitter(this.x + x, this.y + y, direction)
    }

    override val name: String
        get() = "splitter"

    override val visualizationCharacter: Char
        get() = 'S'

    override val blueprintXOffset: Double
        get() = if (direction === UP || direction === DOWN) 0.5 else 0.0
    override val blueprintYOffset: Double
        get() = if (direction === LEFT || direction === RIGHT) 0.5 else 0.0

    override fun addCustomPropertiesToJson(json: JSONObject) {
        json.put("direction", direction.reversed.directionValue)
    }

    override fun rotateCW(around: Int2, count: Int): Building {
        val newDirection = direction.rotateCW(count)

        val newPosition = position.rotateCW(around, count) + when (direction) {
            UP, RIGHT -> when (newDirection) {
                DOWN -> Int2(-1, 0)
                LEFT -> Int2(0, -1)
                else -> Int2(0, 0)
            }
            DOWN, LEFT -> when (newDirection) {
                UP -> Int2(-1, 0)
                RIGHT -> Int2(0, -1)
                else -> Int2(0, 0)
            }
        }

        return Splitter(newPosition.x, newPosition.y, newDirection)
    }
}
